File: brainbox/src/brainbox/loop_predicate.py
"""JMESPath helpers for the Loop runner — Python mirror of app/loop_predicate.go.

Five sites in the runner share this entry point: edge predicates, the
convergence predicate, stop conditions, escalation predicates, and join
conditions (all bool); plus the per-iteration convergence metric (number).
One language, one debugger, one trace line per evaluation across both Go and
Python sides.

Coercion rules MUST match the Go side exactly. The cross-language contract
is exercised by tests/test_loop_predicate.py which uses the same expressions
and envelopes as app/loop_predicate_test.go.
"""

from __future__ import annotations

from typing import Any

import jmespath


def eval_predicate(envelope: Any, expr: str) -> bool:
    """Evaluate ``expr`` against ``envelope`` and coerce to bool.

    Coercion table (matches Go side):
      - True / False             → as-is
      - None                     → False
      - numbers                  → != 0
      - strings                  → non-empty
      - lists / dicts            → non-empty
      - anything else            → raises ValueError

    Empty ``expr`` is always True — same semantics as an Edge with no
    Predicate set.
    """
    if not expr:
        return True
    value = _eval_jmespath(envelope, expr)
    if value is None:
        return False
    if isinstance(value, bool):
        return value
    if isinstance(value, (int, float)):
        return value != 0
    if isinstance(value, str):
        return value != ""
    if isinstance(value, (list, dict)):
        return len(value) > 0
    raise ValueError(f"predicate {expr!r} returned uncoercible type {type(value).__name__}")


def eval_metric(envelope: Any, expr: str) -> float:
    """Evaluate ``expr`` against ``envelope`` and coerce to float.

    Used for the per-iteration convergence metric and any future numeric stop
    conditions (cost_usd, diff_lines, etc.).

    Missing-field returns 0.0 with no error — an iteration that hasn't
    populated findings yet should report a metric of 0, not crash. The
    alternative (treat missing as error) would force every chart query to
    guard for "iteration 0 hasn't run yet."

    Empty ``expr`` is an error — unlike predicates, a metric has no
    sensible default.
    """
    if not expr:
        raise ValueError("empty metric expression")
    value = _eval_jmespath(envelope, expr)
    if value is None:
        return 0.0
    if isinstance(value, bool):
        return 1.0 if value else 0.0
    if isinstance(value, (int, float)):
        return float(value)
    raise ValueError(
        f"metric {expr!r} returned non-numeric type {type(value).__name__}"
    )


def _eval_jmespath(envelope: Any, expr: str) -> Any:
    """Run the expression against the envelope, normalizing through the
    pydantic .model_dump() / .__dict__ surface so json field names drive
    lookups (matching what predicates author against).
    """
    data = _normalize(envelope)
    try:
        return jmespath.search(expr, data)
    except jmespath.exceptions.JMESPathError as exc:
        raise ValueError(f"evaluate {expr!r}: {exc}") from exc


def _normalize(envelope: Any) -> Any:
    """Convert a pydantic model / dataclass / dict into a plain dict/list
    structure jmespath can walk. Leaves scalars and built-in containers alone.
    """
    if envelope is None or isinstance(envelope, (str, int, float, bool, list, dict)):
        return envelope
    if hasattr(envelope, "model_dump"):
        return envelope.model_dump(by_alias=True)
    if hasattr(envelope, "__dict__"):
        return dict(envelope.__dict__)
    return envelope

File: app/loop_predicate.go
package main

import (
	"encoding/json"
	"fmt"

	"github.com/jmespath/go-jmespath"
)

// evalPredicate evaluates a JMESPath expression against an envelope (or any
// JSON-marshallable value) and coerces the result to a bool. Used for edge
// predicates, the convergence predicate, stop conditions, escalation
// predicates, and join conditions — every bool-valued predicate site in the
// Loop runner shares this one entry point.
//
// Coercion rules:
//
//   - true / false → as-is
//   - nil          → false
//   - numbers      → !=0 is true
//   - strings      → non-empty is true (matches operator intuition for things
//                    like `observations.ci_status` returning "green" or "")
//   - arrays/maps  → non-empty is true (matches `findings.blockers` style
//                    expressions where a blockers array existing-but-empty
//                    should still be false-y for "any blockers")
//
// An empty expression is treated as always-true; this matches the semantics
// of an Edge with no Predicate.
func evalPredicate(envelope any, expr string) (bool, error) {
	if expr == "" {
		return true, nil
	}
	v, err := evalJMESPath(envelope, expr)
	if err != nil {
		return false, err
	}
	switch x := v.(type) {
	case nil:
		return false, nil
	case bool:
		return x, nil
	case float64:
		return x != 0, nil
	case int:
		return x != 0, nil
	case string:
		return x != "", nil
	case []interface{}:
		return len(x) > 0, nil
	case map[string]interface{}:
		return len(x) > 0, nil
	default:
		return false, fmt.Errorf("predicate %q returned uncoercible type %T", expr, v)
	}
}

// evalMetric evaluates a JMESPath expression and coerces to float64. Used for
// the per-iteration convergence_metric and for any future numeric stop
// conditions (cost_usd, diff_lines, etc.).
//
// A missing field returns 0 with no error — a Loop iterating before any
// findings have populated should report a metric of 0, not crash. The
// alternative (treat missing as error) would make every chart require a
// guard for "iteration 0 hasn't run yet."
func evalMetric(envelope any, expr string) (float64, error) {
	if expr == "" {
		return 0, fmt.Errorf("empty metric expression")
	}
	v, err := evalJMESPath(envelope, expr)
	if err != nil {
		return 0, err
	}
	switch x := v.(type) {
	case nil:
		return 0, nil
	case float64:
		return x, nil
	case int:
		return float64(x), nil
	case bool:
		if x {
			return 1, nil
		}
		return 0, nil
	default:
		return 0, fmt.Errorf("metric %q returned non-numeric type %T", expr, v)
	}
}

// evalJMESPath runs the expression against the envelope. The envelope is
// first round-tripped through JSON so structs with json tags evaluate using
// the wire-format field names (matching what predicates author against).
func evalJMESPath(envelope any, expr string) (any, error) {
	data, err := normalizeForJMESPath(envelope)
	if err != nil {
		return nil, err
	}
	result, err := jmespath.Search(expr, data)
	if err != nil {
		return nil, fmt.Errorf("evaluate %q: %w", expr, err)
	}
	return result, nil
}

// normalizeForJMESPath ensures the input is a JSON-shaped map/slice/scalar
// rather than a Go struct. If it's already a map/slice/scalar, return as-is.
// Otherwise marshal+unmarshal so json tags drive field names.
func normalizeForJMESPath(envelope any) (any, error) {
	switch envelope.(type) {
	case map[string]interface{}, []interface{}, string, float64, int, bool, nil:
		return envelope, nil
	}
	b, err := json.Marshal(envelope)
	if err != nil {
		return nil, fmt.Errorf("marshal envelope: %w", err)
	}
	var out interface{}
	if err := json.Unmarshal(b, &out); err != nil {
		return nil, fmt.Errorf("unmarshal envelope: %w", err)
	}
	return out, nil
}
